crates/ty_project/src/db.rs

```diff
@@ -126,6 +138,59 @@ impl ProjectDatabase {
         )
     }
 
+    /// Deserialize the database from the persistent cache.
+    pub fn deserialize(&mut self, path: &str, metadata: &ProjectMetadata) -> anyhow::Result<()> {
+        // Read from the persistent cache.
+        let contents = match std::fs::read(path) {
+            Ok(contents) => contents,
+            Err(err) if err.kind() == io::ErrorKind::NotFound => return Ok(()),
+            Err(err) => return Err(err.into()),
+        };
+
+        let _guard = ValueSourceGuard::new(ValueSource::Cli, false);
+
+        // Deserialize the database.
+        let mut decoder = bincode::serde::BorrowedSerdeDecoder::from_slice(
+            &contents,
+            bincode::config::standard(),
+            (),
+        );
+
+        <dyn salsa::Database>::deserialize(self, decoder.as_deserializer())?;
+
+        // Sync any deserialized inputs.
+        for file in File::load_all(self) {
+            file.sync(self);
+            self.files.seed(file, self);
+        }
+
+        for root in FileRoot::load_all(self) {
+            self.files.seed_root(root, self);
+        }
+
+        for project in Project::load_all(self) {
+            if project.root(self) == metadata.root() {
+                project.resolve_settings(self)?;
+                self.project = Some(project);
+                break;
+            }
+        }
```

Comment:
What happens if the user moved the `metadata` file (e.g. if they moved it a directory up)? Do we replace the metadata somewhere (by calling update project), or do we use a stale configuration?

I think what would be ideal is if we wouldn't serialize the project settings at all and always load them instead. But that would require some form of hashing to know when they've changed (including which fields) which is out of scope for now